File: androidApp/src/main/java/me/nathanfallet/bdeensisa/features/clubs/ClubView.kt
package me.nathanfallet.bdeensisa.features.clubs

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.Card
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Text
import androidx.compose.material.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.livedata.observeAsState
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import me.nathanfallet.bdeensisa.features.MainViewModel

@Composable
fun ClubView(
    modifier: Modifier = Modifier,
    viewModel: ClubViewModel,
    mainViewModel: MainViewModel
) {

    val user by mainViewModel.getUser().observeAsState()

    val members by viewModel.getMembers().observeAsState()

    LazyColumn(modifier) {
        item {
            TopAppBar(
                title = { Text(text = viewModel.club.name) },
                actions = {
                    if (members?.any { it.userId == user?.id } == true) {
                        Text(
                            text = "Quitter",
                            modifier = Modifier
                                .clickable {
                                    viewModel.leave(mainViewModel.getToken().value)
                                }
                                .padding(16.dp)
                        )
                    }
                }
            )
        }
        item {
            Spacer(modifier = Modifier.height(8.dp))
        }
        item {
            ClubCard(
                club = viewModel.club,
                badgeText = if (user?.cotisant != null && members?.none { it.userId == user?.id } == true) "REJOINDRE" else null,
                badgeColor = MaterialTheme.colors.primary,
                action = {
                    viewModel.join(mainViewModel.getToken().value)
                },
                detailsEnabled = false
            )
        }
        item {
            Text(
                text = "Membres",
                style = MaterialTheme.typography.h6,
                modifier = Modifier
                    .padding(horizontal = 16.dp)
                    .padding(vertical = 8.dp)
            )
        }
        items(members ?: listOf()) {
            Card(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 16.dp)
                    .padding(vertical = 4.dp),
                elevation = 4.dp
            ) {
                Row(
                    modifier = Modifier
                        .padding(16.dp)
                        .fillMaxWidth(),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.SpaceBetween
                ) {
                    Column(
                        modifier = Modifier.weight(1f, fill = false)
                    ) {
                        Text(
                            text = "${it.user?.firstName} ${it.user?.lastName}",
                            fontWeight = FontWeight.Bold
                        )
                        Text(it.user?.description ?: "")
                    }
                    Text(
                        text = if (it.role == "admin") "ADMIN" else "MEMBRE",
                        style = MaterialTheme.typography.caption,
                        color = Color.White,
                        modifier = Modifier
                            .background(
                                if (it.role == "admin") Color.Black
                                else Color(0xFF0BDA51),
                                MaterialTheme.shapes.small
                            )
                            .padding(horizontal = 10.dp, vertical = 6.dp)
                    )
                }
            }
        }
    }

}

File: iosApp/iosApp/Features/Clubs/ClubView.swift
//
//  ClubView.swift
//  BDE
//
//  Created by Nathan Fallet on 13/02/2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import SwiftUI

struct ClubView: View {
    
    @EnvironmentObject var rootViewModel: RootViewModel
    @StateObject var viewModel: ClubViewModel
    
    var body: some View {
        ScrollView {
            VStack(alignment: .leading) {
                ClubCard(
                    club: viewModel.club,
                    badgeText: rootViewModel.user?.cotisant != nil && !viewModel.members.contains(where: { $0.userId == rootViewModel.user?.id }) ? "REJOINDRE" : nil,
                    badgeColor: .accentColor,
                    action: { _ in
                        viewModel.join(token: rootViewModel.token)
                    },
                    detailsEnabled: false
                )
                HStack {
                    Text("Membres")
                        .font(.title)
                    Spacer()
                }
                .padding(.top)
                LazyVGrid(
                    columns: [GridItem(.adaptive(minimum: 300, maximum: 400))],
                    alignment: .leading
                ) {
                    ForEach(viewModel.members, id: \.userId) { membership in
                        HStack {
                            VStack(alignment: .leading) {
                                Text("\(membership.user?.firstName ?? "") \(membership.user?.lastName ?? "")")
                                    .fontWeight(.bold)
                                Text(membership.user?.description_ ?? "")
                            }
                            Spacer()
                            Text(membership.role == "admin" ? "ADMIN" : "MEMBRE")
                                .font(.caption)
                                .padding(.horizontal, 10)
                                .padding(.vertical, 6)
                                .foregroundColor(.white)
                                .background(membership.role == "admin" ? Color.black : Color.green)
                                .cornerRadius(8)
                        }
                        .cardView()
                    }
                }
            }
            .padding()
        }
        .onAppear(perform: viewModel.onAppear)
        .navigationTitle(Text(viewModel.club.name))
        .toolbar {
            if viewModel.members.contains(where: { $0.userId == rootViewModel.user?.id }) {
                Button("Quitter") {
                    viewModel.leave(token: rootViewModel.token)
                }
            }
        }
    }
    
}
